File: Algorithms/729/MyCalendar.java
class MyCalendar {
    private TreeSet<int[]> events;

    public MyCalendar() {
        events = new TreeSet<>((a, b) -> {
            if (a[1] <= b[0])
                return -1;
            else if (a[0] >= b[1])
                return 1;
            else
                return 0;
        });
    }
    
    public boolean book(int start, int end) {
        return events.add(new int[] {start, end});
    }
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * boolean param_1 = obj.book(start,end);
 */

File: Algorithms/729/MyCalendar.py
# 方法一：直接遍历
# 时间复杂度O(n²)，空间复杂度O(n)
class MyCalendar:

    def __init__(self):
        self.booked = []

    def book(self, startTime: int, endTime: int) -> bool:
        if any(s < endTime and startTime < e for s, e in self.booked):
            return False
        self.booked.append((startTime, endTime))
        return True


# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(startTime,endTime)
